### ISSUE TITLE
Can we have a --endtimer argument?

### ISSUE THREAD

**Issue by ryanjgillies** [streamlink/streamlink#2645]
- [X] This is a feature request and I have read the contribution guidelines.

### Description

I frequently like to schedule streamlink via cron or Task Scheduler depending on the platform. Sometimes this can be to output streams to file at certain times, but normally its so I can jump from stream to stream automatically at whatever time I choose.

A method to time out the stream at a given time in order to allow the output file process to gracefully end, or timeout live streams so that the player can then be scheduled to use a different stream, would be very helpful.

### Expected / Actual behavior

Currently I have to schedule an extra task to kill any instances of streamlink before I can start the next stream.

An `--endtimer` argument that could accept a time format value (for example `00:00:01` for seconds, `00:01:00` for minutes and `01:00:00` for hours) and gracefully close the stream, streamlink and player/file write process would make scheduling much easier!

### Additional comments, screenshots, etc.

Unless of course someone has a better way of exiting streams at pre-allocated times, in which case I'm all ears!

**Comment by back-to** [comment]
https://github.com/streamlink/streamlink/issues/68

**Comment by ryanjgillies** [comment]
Thanks for highlighting the previous threads, didn't find them when I searched using my terms.

However it seems (to me) like the feature request was never really addressed.

#68 only provides a bash solution (for linux/mac) or an ffmpeg option - which means actively using ffmpeg instead of the already incuded `-o` option - and this only applies to streams where the intention is output to file, not simply changing between live streams.

#1749 does address the viewing of streams instead of output to file, BUT only in the instance that its an hls stream.

None of these options fully implement the requested feature, and so I'm still left with the longwinded requirement to kill streamlink externally, rather than from streamlink itself.

In short, everytime I schedule something, I need to do it twice (start and stop), not once (start and timeout).

**Comment by beardypig** [comment]
The problem, @ryanjgillies, is that for some streams we do not know how much time has passed exactly.

**Comment by ryanjgillies** [comment]
Hi @beardypig I guess by how much time has passed, you're referring to the stream in instances when lag or a bit of buffering occurs...?

I can appreciate that would be tricky, but my primary thinking was with regards to watching live streams - watch a sitcom on one channel at 6pm then the football on another channel at 8pm (for example).

So in this instance could the script not keep time based on the clock - a simple count of how much time has elapsed rather than how much of the "stream time" has passed? 

Because in the above example my scheduled task needs to kick in at 8pm, regardless of whether or not the 6pm stream has finished.

**Comment by gravyboat** [comment]
@ryanjgillies In that situation can't you just write a wrapper script that runs and checks the time to start a new Streamlink process at 8? Why should Streamlink handle this instead of an application or script that simply uses Streamlink?

**Comment by ryanjgillies** [comment]
@gravyboat its true I could, but it wouldn't be a cross platform "wrapper script" unless I potentially did some tinkering with python and figured it out that way?

I'm sure there a lot of things I could do with a wrapper script that have already been implemented as features to streamlink - hence this request! The streamlink CLI is very handy and easy to use - personally I think its far more longwinded feeding my streamlink commands into a custom wrapper script over multiple schedules than it is to simply append something like `--endtimer "01:30:00"`

Its something I feel would be of benefit to streamlink users, as also shown in previous requests and enthusiasm by members of the community. But if its not of interest to developers and/or unworkable I of course can appreciate that and keep bodging things on whatever system I happen to use it on!

**Comment by beardypig** [comment]
If you're on linux just use `timeout` or `timelimit`, either of those will serve. If you have one thing scheduled after another, and how I understand it you're not running 2 streams at the same time, you could do a `killall streamlink` or something before starting the next one...
